**复习：**在前面我们已经学习了Pandas基础，知道利用Pandas读取csv数据的增删查改，今天我们要学习的就是**探索性数据分析**，主要介绍如何利用Pandas进行排序、算术计算以及计算描述函数describe()的使用。

# 1 第一章：探索性数据分析

#### 开始之前，导入numpy、pandas包和数据

In [2]:
#加载所需的库
import numpy as np
import pandas as pd

In [4]:
df=pd.read_csv('train_chinese.csv')
df.head(3)



乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价   客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500  NaN    S  
1  female  38.0       1       0          PC 17599  71.2833  C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250  NaN    S

### 1.6 了解你的数据吗？
教材《Python for Data Analysis》第五章

#### 1.6.1 任务一：利用Pandas对示例数据进行排序，要求升序

In [31]:
df3 = pd.DataFrame(np.arange(16).reshape((4,4)),index=['a','c','b','e'],columns=[1,3,2,4])
df3

1   3   2   4
a   0   1   2   3
c   4   5   6   7
b   8   9  10  11
e  12  13  14  15

【代码解析】

pd.DataFrame() ：创建一个DataFrame对象 

np.arange(8).reshape((2, 4)) : 生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7

index=['2, 1] ：DataFrame 对象的索引列

columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行

【问题】：大多数时候我们都是想根据列的值来排序,所以将你构建的DataFrame中的数据根据某一列，升序排列

In [32]:
df3.sort_values(by=4,ascending=True)

1   3   2   4
a   0   1   2   3
c   4   5   6   7
b   8   9  10  11
e  12  13  14  15

【思考】通过书本你能说出Pandas对DataFrame数据的其他排序方式吗？

【总结】下面将不同的排序方式做一个总结

1.让行索引升序排序

In [33]:
df3.sort_index()#行排序


1   3   2   4
a   0   1   2   3
b   8   9  10  11
c   4   5   6   7
e  12  13  14  15

2.让列索引升序排序


In [34]:
df3.sort_index(axis=1)


1   2   3   4
a   0   2   1   3
c   4   6   5   7
b   8  10   9  11
e  12  14  13  15

3.让列索引降序排序

In [35]:
df3.sort_index(axis=1, ascending=False)


4   3   2   1
a   3   1   2   0
c   7   5   6   4
b  11   9  10   8
e  15  13  14  12

4.让任选两列数据同时降序排序

In [36]:
df3.sort_values(by=[2, 3], ascending=False)


1   3   2   4
e  12  13  14  15
b   8   9  10  11
c   4   5   6   7
a   0   1   2   3

#### 1.6.2 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从这个数据中你可以分析出什么？

In [40]:
df.sort_values(by=['票价', '年龄'], ascending=False).head(20)


乘客ID  是否幸存  仓位等级                                                 姓名  \
679   680     1     1                 Cardeza, Mr. Thomas Drake Martinez   
258   259     1     1                                   Ward, Miss. Anna   
737   738     1     1                             Lesurer, Mr. Gustave J   
438   439     0     1                                  Fortune, Mr. Mark   
341   342     1     1                     Fortune, Miss. Alice Elizabeth   
88     89     1     1                         Fortune, Miss. Mabel Helen   
27     28     0     1                     Fortune, Mr. Charles Alexander   
742   743     1     1              Ryerson, Miss. Susan Parker "Suzette"   
311   312     1     1                         Ryerson, Miss. Emily Borie   
299   300     1     1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
118   119     0     1                           Baxter, Mr. Quigg Edmond   
380   381     1     1                              Bidois, Miss. Rosalie   
716   717     1     1                      Endres, Miss. Caroline Louise   
700   701     1     1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
557   558     0     1                                Robbins, Mr. Victor   
527   528     0     1                                 Farthing, Mr. John   
377   378     0     1                          Widener, Mr. Harry Elkins   
779   780     1     1  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...   
730   731     1     1                      Allen, Miss. Elisabeth Walton   
689   690     1     1                  Madill, Miss. Georgette Alexandra   

         性别    年龄  兄弟姐妹个数  父母子女个数      船票信息        票价               客舱 登船港口  
679    male  36.0       0       1  PC 17755  512.3292      B51 B53 B55    C  
258  female  35.0       0       0  PC 17755  512.3292              NaN    C  
737    male  35.0       0       0  PC 17755  512.3292             B101    C  
438    male  64.0       1       4     19950  263.0000      C23 C25 C27    S  
341  female  24.0       3       2     19950  263.0000      C23 C25 C27    S  
88   female  23.0       3       2     19950  263.0000      C23 C25 C27    S  
27     male  19.0       3       2     19950  263.0000      C23 C25 C27    S  
742  female  21.0       2       2  PC 17608  262.3750  B57 B59 B63 B66    C  
311  female  18.0       2       2  PC 17608  262.3750  B57 B59 B63 B66    C  
299  female  50.0       0       1  PC 17558  247.5208          B58 B60    C  
118    male  24.0       0       1  PC 17558  247.5208          B58 B60    C  
380  female  42.0       0       0  PC 17757  227.5250              NaN    C  
716  female  38.0       0       0  PC 17757  227.5250              C45    C  
700  female  18.0       1       0  PC 17757  227.5250          C62 C64    C  
557    male   NaN       0       0  PC 17757  227.5250              NaN    C  
527    male   NaN       0       0  PC 17483  221.7792              C95    S  
377    male  27.0       0       2    113503  211.5000              C82    C  
779  female  43.0       0       1     24160  211.3375               B3    S  
730  female  29.0       0       0     24160  211.3375               B5    S  
689  female  15.0       0       1     24160  211.3375               B5    S

In [55]:
df.sort_values(by=['是否幸存', '仓位等级'], ascending=False).head(100)


乘客ID  是否幸存  仓位等级                                                 姓名  \
2       3     1     3                             Heikkinen, Miss. Laina   
8       9     1     3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
10     11     1     3                    Sandstrom, Miss. Marguerite Rut   
19     20     1     3                            Masselmani, Mrs. Fatima   
22     23     1     3                        McGowan, Miss. Anna "Annie"   
..    ...   ...   ...                                                ...   
691   692     1     3                                 Karun, Miss. Manca   
692   693     1     3                                       Lam, Mr. Ali   
697   698     1     3                   Mullens, Miss. Katherine "Katie"   
709   710     1     3  Moubarek, Master. Halim Gonios ("William George")   
727   728     1     3                           Mannion, Miss. Margareth   

         性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价   客舱 登船港口  
2    female  26.0       0       0  STON/O2. 3101282   7.9250  NaN    S  
8    female  27.0       0       2            347742  11.1333  NaN    S  
10   female   4.0       1       1           PP 9549  16.7000   G6    S  
19   female   NaN       0       0              2649   7.2250  NaN    C  
22   female  15.0       0       0            330923   8.0292  NaN    Q  
..      ...   ...     ...     ...               ...      ...  ...  ...  
691  female   4.0       0       1            349256  13.4167  NaN    C  
692    male   NaN       0       0              1601  56.4958  NaN    S  
697  female   NaN       0       0             35852   7.7333  NaN    Q  
709    male   NaN       1       1              2661  15.2458  NaN    C  
727  female   NaN       0       0             36866   7.7375  NaN    Q  

[100 rows x 12 columns]

【思考】排序后，如果我们仅仅关注年龄和票价两列。根据常识我知道发现票价越高的应该客舱越好，所以我们会明显看出，票价前20的乘客中存活的有14人，这是相当高的一个比例，那么我们后面是不是可以进一步分析一下票价和存活之间的关系，年龄和存活之间的关系呢？当你开始发现数据之间的关系了，数据分析就开始了。

当然，这只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

**多做几个数据的排序**

In [4]:
#代码


In [4]:
#写下你的思考






#### 1.6.3 任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果

In [48]:

frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])
frame1_a

a    b    c
one    0.0  1.0  2.0
two    3.0  4.0  5.0
three  6.0  7.0  8.0

In [49]:
frame1_b


a     e     c
first   0.0   1.0   2.0
one     3.0   4.0   5.0
two     6.0   7.0   8.0
second  9.0  10.0  11.0

将frame_a和frame_b进行相加


In [50]:
frame1_a + frame1_b


a   b     c   e
first   NaN NaN   NaN NaN
one     3.0 NaN   7.0 NaN
second  NaN NaN   NaN NaN
three   NaN NaN   NaN NaN
two     9.0 NaN  13.0 NaN

【提醒】两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN。<br>
当然，DataFrame还有很多算术运算，如减法，除法等，有兴趣的同学可以看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分，多在网络上查找相关学习资料。

#### 1.6.4 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [54]:
'''
还是用之前导入的chinese_train.csv如果我们想看看在船上，最大的家族有多少人（‘兄弟姐妹个数’+‘父母子女个数’），我们该怎么做呢？
'''
max(df['兄弟姐妹个数'] + df['父母子女个数'])

10

In [4]:
#代码


【提醒】我们只需找出”兄弟姐妹个数“和”父母子女个数“之和最大的数，当然你还可以想出很多方法和思考角度，欢迎你来说出你的看法。

**多做几个数据的相加，看看你能分析出什么？**

In [4]:
#代码


In [5]:
#写下你的其他分析







#### 1.6.5 任务五：学会使用Pandas describe()函数查看数据基本统计信息

In [56]:
#(1) 关键知识点示例做一遍（简单数据）
# 具体请看《利用Python进行数据分析》第五章 汇总和计算描述统计 部分

#自己构建一个有数字有空值的DataFrame数据



frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [4]:
#代码


调用 describe 函数，观察frame2的数据基本信息

In [57]:
frame2.describe()


one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

#### 1.6.6 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？

In [58]:
'''
看看泰坦尼克号数据集中 票价 这列数据的基本统计数据
'''
df['票价'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: 票价, dtype: float64

上面数据我们可以看出， 一共有891个票价数据， 平均值约为：32.20， 标准差约为49.69，说明票价波动特别大， 25%的人的票价是低于7.91的，50%的人的票价低于14.45，75%的人的票价低于31.00， 票价最大值约为512.33，最小值为0。


【思考】从上面数据我们可以看出，试试在下面写出你的看法。然后看看我们给出的答案。

当然，答案只是我的想法，你还可以有更多想法，欢迎写在你的学习笔记中。

**多做几个组数据的统计，看看你能分析出什么？**

In [6]:
# 写下你的其他分析




【思考】有更多想法，欢迎写在你的学习笔记中。

【总结】本节中我们通过Pandas的一些内置函数对数据进行了初步统计查看，这个过程最重要的不是大家得掌握这些函数，而是看懂从这些函数出来的数据，构建自己的数据分析思维，这也是第一章最重要的点，希望大家学完第一章能对数据有个基本认识，了解自己在做什么，为什么这么做，后面的章节我们将开始对数据进行清洗，进一步分析。